library/datasort.c

```diff
@@ -560,6 +561,72 @@ static int datasort_split_iterator_free(struct eblob_iterate_control *ictl __att
 	return 0;
 }
 
+/*
+ * Create a view over a blob and insert it in dcfg->unsorted_list
+ * Assume that blob has already sorted
+ */
+static int datasort_add_view_chunk(struct datasort_cfg *dcfg, int index) {
+	assert(dcfg);
+	assert(index >= 0 && index < dcfg->bctl_cnt);
+
+	int err = 0;
+	struct eblob_base_ctl *bctl = dcfg->bctl[index];
+	assert(bctl);
+
+	EBLOB_WARNX(dcfg->log, EBLOB_LOG_INFO, "defrag: datasort_add_view_chunk: start, name: %s",
+			dcfg->bctl[index]->name);
+
+	struct datasort_chunk *chunk = calloc(1, sizeof(*chunk));
+	if (chunk == NULL) {
+		EBLOB_WARNC(dcfg->log, EBLOB_LOG_ERROR, errno, "defrag: add_view_chunk: can't allocate memory for datasort_chunk");
+		return -errno;
+	}
+
+	eblob_base_wait_locked(bctl);
+	err = eblob_base_setup_data(bctl, 0);
+	if (err) {
+		pthread_mutex_unlock(&bctl->lock);
+                goto err_out;
+	}
+
+	struct eblob_file_ctl *index_ctl = &bctl->index_ctl;
+	assert(index_ctl->fd > 0);
+
+	chunk->fd = bctl->data_ctl.fd;
+	chunk->already_sorted = 1;
+	chunk->base_view = 1;
+	chunk->offset = bctl->data_ctl.offset;
+	chunk->index_size = index_ctl->size;
+	chunk->count = chunk->index_size / sizeof(struct eblob_disk_control);
+	chunk->index = calloc(chunk->count, sizeof(struct eblob_disk_control));
+	if (chunk->index == NULL) {
+		EBLOB_WARNX(dcfg->log, EBLOB_LOG_ERROR, "defrag: load_index: can't allocate memory for index of size %zd",
+			chunk->count * sizeof(struct eblob_disk_control));
+		err = -errno;
+		goto err_out;
```

Comment:
Don't forget about `pthread_mutex_unlock(&bctl->lock);`. Better to use a cascade of error labels to go to:
```
err_out_with_lock:
   unlock(lock);
err_out:
   some_actions
}
```